<a href="https://colab.research.google.com/github/wux8es/EDA/blob/main/Copy_of_assignment2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment: Data Wrangling and Exploratory Data Analysis
## Do Q1 and Q2, and one other question.

In [1]:
! git clone https://www.github.com/wux8es/assignment2

Cloning into 'assignment2'...
remote: Enumerating objects: 17, done.
remote: Counting objects: 100% (6/6), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 17 (delta 2), reused 2 (delta 2), pack-reused 11
Receiving objects: 100% (17/17), 5.16 MiB | 10.51 MiB/s, done.
Resolving deltas: 100% (2/2), done.


**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

1) The paper is about "data tidying," a way to make data cleaning easy and effective.
2) It is intended to a provide a standard way to organize data without needing to "start from scratch" every time.
3) There are unique ways that data sets need to be cleaned. For each case, it is easy to define what are variables and observations. For the whole data set, it is more difficult.
4) Values are numbers or strings organized into a data set. Each value belongs to a variable and an observation. A variable contains all measurments of one attribute (like the heights of multiple people). An observation is all values measured on one unit (for example, one person).
5) Each variable forms a column, each observation forms a row, and each type of observational unit forms a table.
6) Column headers are values, not variable names. Multiple variables are stored in one column. Variables are stored in both rows and columns. Multiple types of observational units are stored in the same table. A single observational unit is stored in multiple tables. The columns need to be turned into rows. Melting means to reformat the data set turning the columns into rows.
7) Because table 11 has a column for each day of the month, and is not easily comprehensible. Also, the observations are columns and should be rows. Table 12(a) is molten because the columns and rows are switched appropriately. Table 12(b) is tidy because it includes variables instead of names in the third column.
8) "If tidy data is only as useful as the tools that work with it, then tidy tools will be inextricably linked to tiny data." Wickham hopes others will use improve this framework to develop beter stroage strategies and tools.

**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_NYC.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [ ]:
# Part 1
import pandas as pd
import numpy as np
df=pd.read_csv('./assignment2/data/airbnb_NYC.csv', encoding='latin1')
# Remove commas
df['Price'] = df['Price'].str.replace(',','')
# Convert strings to integers
df['Price'] = [int(each) for each in df['Price']]
# I found no missing variables, looked using unique()

In [ ]:
# Part 2
import pandas as pd
import numpy as np
df=pd.read_csv('./assignment2/data/sharks.csv', low_memory = False)
# Looking for missing function using unqiue
print(df['Type'].unique())
# Making map for cleaning (ChatGPT helped). I am merging synonyms with eachother
cleaning_map = {
    'Boating': 'Watercraft',
    'Boat': 'Watercraft',
    'Boatomg': 'Watercraft',
    'Sea Disaster': 'Disaster',
    'Unverified': 'Unconfirmed'}
# Rename values
df['Type'] = df['Type'].replace(cleaning_map)
print(df['Type'].unique())
# Missing values are already market as nan, so I am ok not renaming them

['Unprovoked' 'Provoked' 'Questionable' 'Watercraft' 'Unconfirmed'
 'Unverified' 'Invalid' 'Under investigation' 'Boating' 'Sea Disaster' nan
 'Boat' 'Boatomg']
['Unprovoked' 'Provoked' 'Questionable' 'Watercraft' 'Unconfirmed'
 'Invalid' 'Under investigation' 'Disaster' nan]


In [33]:
# Part 3
import pandas as pd
import numpy as np
# Downloaded the file and uploaded from computer. I was having trouble consistently accessing the URL.
df = pd.read_csv('/content/October 2017 Cohort_Virginia Pretrial Data Project_Deidentified FINAL Update_10272021 (2) (1).csv',low_memory=False)
print(df['WhetherDefendantWasReleasedPretrial'].unique())
# Assuming that 0 means 'no' and 1 means 'yes,' I will consider 9 a missing value and replace it with np.nan
df['WhetherDefendantWasReleasedPretrial'] = df['WhetherDefendantWasReleasedPretrial'].replace(9, 'np.nan')
print(df['WhetherDefendantWasReleasedPretrial'].unique())



[9 0 1]
['np.nan' 0 1]


In [46]:
# Part 4
import pandas as pd
import numpy as np
# Comparing these two columns, it seems that the 'ImposedSentenceAllChargeInContactEvent' column should also have whole numbers
# I rounded all numbers up and replaced missing values with np.nan
df['ImposedSentenceAllChargeInContactEvent'] = df['ImposedSentenceAllChargeInContactEvent'].replace(' ', 'np.nan')

def convert_to_int(value):
    try:
        return int(float(value))
    except ValueError:
        return value
df['ImposedSentenceAllChargeInContactEvent'] = df['ImposedSentenceAllChargeInContactEvent'].apply(convert_to_int)
print(df['ImposedSentenceAllChargeInContactEvent'].unique())

# Chat GPT helped.



['np.nan' 60 12 0 36 6 24 5 120 72 11 2 84 108 300 240 180 4 96 54 44 115
 132 48 258 34 76 111 9 3 1 35 106 8 18 480 32 93 234 732 21 7 600 43 179
 52 30 20 192 702 14 55 53 114 68 46 102 65 200 57 117 81 22 1980 56 10 47
 1500 40 284 118 42 162 156 105 51 246 29 75 324 360 59 66 660 78 228 62
 86 168 23 33 720 348 1200 27 49 87 420 63 79 17 238 126 45 158 216 227
 445 70 516 177 1752 90 1080 141 230 31 2208 69 26 140 131 344 219 101 71
 58 67 91 155 92 207 74 518 28 237 404 551 39 15 124 176 19 482 129 88 128
 136 50 363 288 250 107 444 205 38 264 276 173 222 144 294 336 431 450 73
 99 127 202 89 242 1029 194 399 198 148 13 104 212 183 252 540 133 456 103
 576 492 61 16 138 95 77 1011 188 432 98 112 354 123 204 197 174 840 440
 172 146 64 312 82 121 41 473 109 94 302 154 83 110 226 37 286 283 165 175
 232 152 167 119 960 116]


**Q3.** This question provides some practice doing exploratory data analysis and visualization.

The "relevant" variables for this question are:
  - `level` - Level of institution (4-year, 2-year)
  - `aid_value` - The average amount of student aid going to undergraduate recipients
  - `control` - Public, Private not-for-profit, Private for-profit
  - `grad_100_value` - percentage of first-time, full-time, degree-seeking undergraduates who complete a degree or certificate program within 100 percent of expected time (bachelor's-seeking group at 4-year institutions)

1. Load the `./data/college_completion.csv` data with Pandas.
2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
3. Cross tabulate `control` and `level`. Describe the patterns you see.
4. For `grad_100_value`, create a histogram, kernel density plot, boxplot, and statistical description.
5. For `grad_100_value`, create a grouped kernel density plot by `control` and by `level`. Describe what you see. Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `grad_100_value` by `level` and `control`. Which institutions appear to have the best graduation rates?
6. Create a new variable, `df['levelXcontrol']=df['level']+', '+df['control']` that interacts level and control. Make a grouped kernel density plot. Which institutions appear to have the best graduation rates?
7. Make a kernel density plot of `aid_value`. Notice that your graph is "bi-modal", having two little peaks that represent locally most common values. Now group your graph by `level` and `control`. What explains the bi-modal nature of the graph? Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `aid_value` by `level` and `control`.
8. Make a scatterplot of `grad_100_value` by `aid_value`. Describe what you see. Now make the same plot, grouping by `level` and then `control`. Describe what you see. For which kinds of institutions does aid seem to increase graduation rates?

**Q4.** This question uses the Airbnb data to practice making visualizations.

  1. Load the `./data/airbnb_NYC.csv` data with Pandas. You should have cleaned the `Price` variable in question 2, and you'll need it later for this question.
  2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
  3. Cross tabulate `Room Type` and `Property Type`. What patterns do you see in what kinds of rentals are available? For which kinds of properties are private rooms more common than renting the entire property?
  4. For `Price`, make a histogram, kernel density, box plot, and a statistical description of the variable. Are the data badly scaled? Are there many outliers? Use `log` to transform price into a new variable, `price_log`, and take these steps again.
  5. Make a scatterplot of `price_log` and `Beds`. Describe what you see. Use `.groupby()` to compute a desciption of `Price` conditional on/grouped by the number of beds. Describe any patterns you see in the average price and standard deviation in prices.
  6. Make a scatterplot of `price_log` and `Beds`, but color the graph by `Room Type` and `Property Type`. What patterns do you see? Compute a description of `Price` conditional on `Room Type` and `Property Type`. Which Room Type and Property Type have the highest prices on average? Which have the highest standard deviation? Does the mean or median appear to be a more reliable estimate of central tendency, and explain why?
  7. We've looked a bit at this `price_log` and `Beds` scatterplot. Use seaborn to make a `jointplot` with `kind=hex`. Where are the data actually distributed? How does it affect the way you think about the plots in 5 and 6?

**Q5.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?

1. The census asked through 5 categories (Black, White, American Indian/Alaska Native, Asian, Native Hawaiian or other Pacific Islander). Respondents can also designate themselves under the “Some Other Race” category and can report more than one race.
2. We gather these data to minimize discrimination and ensure fair treatment. These data can encourage legislation and action for a society with equal opportunity. The effectiveness and fairness of the legislation is dependent on the quality of the data.
3. It is good we are collecting data on race, but there should be more more options when designating yourself as a specific race instead of just “Some Other Race. Within practical reason, I believe the more specific the census gets about race/gender data, the better.
4. According to the census website, the census only asks about current sex and does not ask about gender or sexual orientation. It may be helpful to also ask about gender/sexual orientation to gather data on gender/sexual orientation trends and associations in the U.S.
5. There may be many observations, making it it difficult to clean and organize the data. Moral judgements must be made to handle missing values. Data scientists must decide what can truly be designated as missing values, and what must be kept. Data scientists may become careless about lumping everyone’s gender and sexual characteristics together.
6. My main concern with an algorithm like this is its purpose. Why would someone need to impute such characteristics? What are they used for? Collecting such data can have grave consequences as we have seen throughout history.


**Q6.** Open the `./data/CBO_data.pdf` file. This contains tax data for 2019, explaining where the money comes from that the U.S. Federal Government Spends in terms of taxation on individuals/families and payroll taxes (the amount that your employer pays in taxes on your wages).

For some context, the Federal government ultimately spent about $4.4 trillion in 2019, which was 21% of GDP (the total monetary value of all goods and services produced within the United States). Individual Income Taxes is the amount individuals pay on their wages to the Federal government, Corporate Income Taxes is the taxes individuals pay on capital gains from investment when they sell stock or other financial instruments, Payroll Taxes is the tax your employer pays on your wages, Excises and Customs Duties are taxes on goods or services like sin taxes on cigarettes or alcohol, and Estate and Gift Taxes are taxes paid on transfers of wealth to other people.

1. Get the Millions of Families and Billions of Dollars data into a .csv file and load it with Pandas.
2. Create a bar plot of individual income taxes by income decile. Explain what the graph shows. Why are some values negative?
3. Create a bar plot of Total Federal Taxes by income decile. Which deciles are paying net positive amounts, and which are paying net negative amounts?
4. Create a stacked bar plot for which Total Federal Taxes is grouped by Individual Income Taxes, Payroll Taxes, Excises and Customs Duties, and Estate and Gift Taxes. How does the share of taxes paid vary across the adjusted income deciles? (Hint: Are these the kind of data you want to melt?)
5. Below the Total line for Millions of Families and Billions of Dollars, there are data for the richest of the richest families. Plot this alongside the bars for the deciles above the Total line. Describe your results.
6. Get the Percent Distribution data into a .csv file and load it with Pandas. Create a bar graph of Total Federal Taxes by income decile.
7. A tax system is progressive if higher-income and wealthier individuals pay more than lower-income and less wealthy individuals, and it is regressive if the opposite is true. Is the U.S. tax system progressive in terms of amount paid? In terms of the percentage of the overall total?
8. Do the rich pay enough in taxes? Defend your answer.